File: src/main/java/org/adoptopenjdk/jheappo/heapdump/EncodedChunk.java
package org.adoptopenjdk.jheappo.heapdump;

import java.io.PrintStream;
import org.adoptopenjdk.jheappo.model.BasicDataTypeValue;
import org.adoptopenjdk.jheappo.objects.BasicDataTypes;

public class EncodedChunk {
    private final byte[] body;
    private int index;

    public EncodedChunk(byte[] body) {
        this(body, 0);
    }

    private EncodedChunk(byte[] body, int index) {
        this.body = body;
        this.index = index;
    }

    public byte[] read(int bufferLength) {
        byte[] buffer = new byte[bufferLength];
        for (int i = 0; i < bufferLength; i++) {
            buffer[i] = (byte) (body[index++] & 0xff);
        }
        return buffer;
    }

    protected byte[] getBody() {
        return body;
    }
    public void skip(int skipOver) {
        index += skipOver;
    }
    public byte[] readRemaining() {
        return read( body.length - index);
    }
    public int getIndex() { return index; }

    public boolean endOfBuffer() {  return getBody().length <= index; }

    private int read() { return body[index++] & 0xff; }

    public byte extractU1() {
        return (byte)read();
    }

    public short extractU2() {
        int value = extractU1();
        for (int cursor = 1; cursor < 2; cursor++) {
            value = (value << 8) | (short)read();
        }
        return (short)value;
    }

    public int extractU4() {
        int value = extractU1();
        for (int cursor = 1; cursor < 4; cursor++) {
            value = (value << 8) | read();
        }
        return value;
    }

    public long extractU8() {
        long value = extractU1();
        for (int cursor = 1; cursor < 8; cursor++) {
            value = (value << 8) | (long)read();
        }
        return value;
    }

    public long extractID() {
        return extractU8();
    }

    public boolean extractBoolean() {
        return extractU1() != 0;
    }

    public char extractChar() {
        return (char)extractU2();
    }

    public byte extractByte() {
        return extractU1();
    }

    public short extractShort() {
        return extractU2();
    }

    public float extractFloat() {
        return Float.intBitsToFloat(extractInt());
    }

    public int extractInt() {
        return extractU4();
    }

    public double extractDouble() {
        return Double.longBitsToDouble(extractLong());
    }

    public long extractLong() {
        return extractU8();
    }

    public BasicDataTypeValue extractBasicType(BasicDataTypes basicType) {

        switch (basicType) {
            case BOOLEAN :
                return new BasicDataTypeValue(this.extractBoolean(),BasicDataTypes.BOOLEAN);
            case CHAR :
                return new BasicDataTypeValue(this.extractChar(),BasicDataTypes.CHAR);
            case BYTE :
                return new BasicDataTypeValue(this.extractByte(),BasicDataTypes.BYTE);
            case SHORT:
                return new BasicDataTypeValue(this.extractShort(),BasicDataTypes.SHORT);
            case FLOAT :
                return new BasicDataTypeValue(this.extractFloat(),BasicDataTypes.FLOAT);
            case INT :
                return new BasicDataTypeValue(this.extractInt(),BasicDataTypes.INT);
            case OBJECT :
                return new BasicDataTypeValue(this.extractID(),BasicDataTypes.OBJECT);
            case DOUBLE :
                return new BasicDataTypeValue(this.extractDouble(),BasicDataTypes.DOUBLE);
            case LONG:
                return new BasicDataTypeValue(this.extractLong(),BasicDataTypes.LONG);
            default:
                return new BasicDataTypeValue(null,BasicDataTypes.UNKNOWN);
        }
    }

    public BasicDataTypeValue extractBasicType(int basicType) {
        return extractBasicType(BasicDataTypes.fromInt(basicType));
    }

    /**
     * Useful when you want to read some bytes without affecting the internal cursor position.
     *
     * @return A chunk with the same byte array and offset as this object.
     */
    public EncodedChunk copy() {
        return new EncodedChunk(body, index);
    }

    /*
    Debugging aid
     */
    public void dump(PrintStream out) {
        int max = (body.length > 1000) ? 1000 : body.length;
        for( int cursor = 0; cursor < max; cursor++) {
            System.out.print(Integer.toHexString(body[cursor] & 255));
            System.out.print( " ");
        }
        out.println("");
        for( int cursor = 0; cursor < max; cursor++) {
            System.out.print((char)(body[cursor] & 255));
            System.out.print( " ");
        }
        out.println("");
    }

}

File: src/main/java/org/adoptopenjdk/jheappo/heapdump/EncodedChunk.kt
package org.adoptopenjdk.jheappo.heapdump

import java.io.PrintStream
import org.adoptopenjdk.jheappo.model.BasicDataTypeValue
import org.adoptopenjdk.jheappo.objects.BasicDataTypes
import kotlin.experimental.and

/**
 * A wrapper around bytes that represents the encoding used by the hprof binary format.
 *
 * This is used by various hprof record types representing stack frames, objects in the heap, etc to decode the
 * corresponding components (e.g. id numbers). Notably, most things are represented as unsigned ints, which we cannot
 * represent well (yet).
 *
 * See https://hg.openjdk.java.net/jdk/jdk/file/9a73a4e4011f/src/hotspot/share/services/heapDumper.cpp for details on
 * the format.
 */
class EncodedChunk private constructor(private val body: ByteArray, index: Int) {
    var index: Int = 0
        private set

    constructor(body: ByteArray) : this(body, 0) {}

    init {
        this.index = index
    }

    fun read(bufferLength: Int): ByteArray {
        val buffer = ByteArray(bufferLength)
        for (i in 0 until bufferLength) {
            buffer[i] = (body[index++].toInt() and 0xff).toByte()
        }
        return buffer
    }

    fun skip(skipOver: Int) {
        index += skipOver
    }

    fun readRemaining(): ByteArray {
        return read(body.size - index)
    }

    fun endOfBuffer(): Boolean {
        return body.size <= index
    }

    private fun read(): Int {
        return body[index++].toInt() and 0xff
    }

    fun extractU1(): Byte {
        return read().toByte()
    }

    fun extractU2(): Short {
        var value = extractU1().toInt()
        for (cursor in 1..1) {
            value = value shl 8 or read()
        }
        return value.toShort()
    }

    fun extractU4(): Int {
        var value = extractU1().toInt()
        for (cursor in 1..3) {
            value = value shl 8 or read()
        }
        return value
    }

    fun extractU8(): Long {
        var value = extractU1().toLong()
        for (cursor in 1..7) {
            value = value shl 8 or read().toLong()
        }
        return value
    }

    fun extractID(): Long {
        return extractU8()
    }

    fun extractBoolean(): Boolean {
        return extractU1().toInt() != 0
    }

    fun extractChar(): Char {
        return extractU2().toChar()
    }

    fun extractByte(): Byte {
        return extractU1()
    }

    fun extractShort(): Short {
        return extractU2()
    }

    fun extractFloat(): Float {
        return java.lang.Float.intBitsToFloat(extractInt())
    }

    fun extractInt(): Int {
        return extractU4()
    }

    fun extractDouble(): Double {
        return java.lang.Double.longBitsToDouble(extractLong())
    }

    fun extractLong(): Long {
        return extractU8()
    }

    fun extractBasicType(basicType: BasicDataTypes): BasicDataTypeValue {
        return when (basicType) {
            BasicDataTypes.BOOLEAN -> BasicDataTypeValue(this.extractBoolean(), BasicDataTypes.BOOLEAN)
            BasicDataTypes.CHAR -> BasicDataTypeValue(this.extractChar(), BasicDataTypes.CHAR)
            BasicDataTypes.BYTE -> BasicDataTypeValue(this.extractByte(), BasicDataTypes.BYTE)
            BasicDataTypes.SHORT -> BasicDataTypeValue(this.extractShort(), BasicDataTypes.SHORT)
            BasicDataTypes.FLOAT -> BasicDataTypeValue(this.extractFloat(), BasicDataTypes.FLOAT)
            BasicDataTypes.INT -> BasicDataTypeValue(this.extractInt(), BasicDataTypes.INT)
            BasicDataTypes.OBJECT -> BasicDataTypeValue(this.extractID(), BasicDataTypes.OBJECT)
            BasicDataTypes.DOUBLE -> BasicDataTypeValue(this.extractDouble(), BasicDataTypes.DOUBLE)
            BasicDataTypes.LONG -> BasicDataTypeValue(this.extractLong(), BasicDataTypes.LONG)
            else -> BasicDataTypeValue(null, BasicDataTypes.UNKNOWN)
        }
    }

    fun extractBasicType(basicType: Int): BasicDataTypeValue {
        return extractBasicType(BasicDataTypes.fromInt(basicType)!!)
    }

    /**
     * Useful when you want to read some bytes without affecting the internal cursor position.
     *
     * @return A chunk with the same byte array and offset as this object.
     */
    fun copy(): EncodedChunk {
        return EncodedChunk(body, index)
    }

    /*
    Debugging aid
     */
    fun dump(out: PrintStream) {
        val max = if (body.size > 1000) 1000 else body.size
        for (cursor in 0 until max) {
            print(Integer.toHexString(body[cursor].toInt() and 255))
            print(" ")
        }
        out.println("")
        for (cursor in 0 until max) {
            print((body[cursor].toInt() and 255).toChar())
            print(" ")
        }
        out.println("")
    }

}
